### PR TITLE
soc: arm: stm32h7: rework STM32H7 dual core boot

### LINKED ISSUE #38069: stm32h747i_disco M4 not working following merge of 9fa5437447712eece9c88e728ac05ac10fb01c4a
**Describe the bug**
Since merge of 9fa5437447712eece9c88e728ac05ac10fb01c4a, blinky sample ca'nt work on stm32h747i_disco_m4 target

**To Reproduce**
cd zephyr
west build -p -b stm32h747i_disco_m4 samples/basic/blinky
west flash

**Expected behavior**
Sample should work.

**Environment (please complete the following information):**
 - OS: Linux
 - SDK 0.13.0
 - SHA1: Starting 9fa5437447712eece9c88e728ac05ac10fb01c4a

**Additional context**
Both procedures below have a positive impact:
- Run following command on m7 core: `west debug` `continue` > Blink starts
- Disable MPU and SERIAL (CONFIG_ARM_MPU=n, CONFIG_SERIAL=n) > Blink starts

Potentially linked to https://github.com/zephyrproject-rtos/zephyr/issues/37827.

Despite issue appears after 9fa5437447712eece9c88e728ac05ac10fb01c4a, I don't think this change has a direct impact but rather reveals a already existing, silent, issue.

### PR DESCRIPTION
**Pull request by ABOSTM**
soc: arm: stm32h7: rework STM32H7 dual core boot

It happens that CM7 wakeups CM4, before CM4 goes to sleep.
Thus when CM4 goes to sleep,
there no more wakeup from CM7. And CM4 hangs.
For a simple synchronisation implementation,
CM4 doesn't go to sleep any more,
instead it waits (active wait) for CM7 to take HSEM
(meaning that clock configuration is finished).

Fixes #38069

### CODE REVIEW

**Comment by ABOSTM** [comment]
@cameled, This should fixes the issue you mentioned in #38322

**Comment by cameled** [comment]
> @cameled, This should fixes the issue you mentioned in #38322

I tested the #38322 with this PR(BCM4 enabled) .  open-amp sample still not working on CM4.

Below is my test  result:
```
  CM7:
     open-amp sample  -- working

  CM4:
    1. open-amp sample  -- not working
    2. blinky  ---working
```

**Comment by erwango** [comment]
^^@cameled

**Comment by erwango** [comment]
> ow is my test result:
> 
> ```
>   CM7:
>      open-amp sample  -- working
> 
>   CM4:
>     1. open-amp sample  -- not working
>     2. blinky  ---working
> ```

Thanks for the tests and feedback. Though we should keep in might this may be related to other issues.
What are openamp test result when BCM4 disabled ?

**Comment by cameled** [comment]
> > ow is my test result:
> > ```
> >   CM7:
> >      open-amp sample  -- working
> > 
> >   CM4:
> >     1. open-amp sample  -- not working
> >     2. blinky  ---working
> > ```
> 
> Thanks for the tests and feedback. Though we should keep in might this may be related to other issues.
> What are openamp test result when BCM4 disabled ?

openamp work fine on CM4 when BCM4 disabled.

File: soc/arm/st_stm32/stm32h7/soc_m4.c [hunk 1]
```diff
@@ -81,21 +48,20 @@ static int stm32h7_m4_init(const struct device *arg)
 
 	irq_unlock(key);
 
-	/*HW semaphore Clock enable*/
-	LL_AHB4_GRP1_EnableClock(LL_AHB4_GRP1_PERIPH_HSEM);
-
 	/* In case CM4 has not been forced boot by CM7,
 	 * CM4 needs to be stopped until CM7 has setup clock configuration
```
Comment (erwango) [inline soc/arm/st_stm32/stm32h7/soc_m4.c:52]: Comment required update

Comment (ABOSTM) [reply soc/arm/st_stm32/stm32h7/soc_m4.c]: Done

**Comment by ABOSTM** [comment]
> > > ow is my test result:
> > > ```
> > >   CM7:
> > >      open-amp sample  -- working
> > > 
> > >   CM4:
> > >     1. open-amp sample  -- not working
> > >     2. blinky  ---working
> > > ```
> > 
> > 
> > Thanks for the tests and feedback. Though we should keep in might this may be related to other issues.
> > What are openamp test result when BCM4 disabled ?
> 
> openamp work fine on CM4 when BCM4 disabled.

@cameled , I tried to reproduce your problem with openamp, but when I compile for stm32h747i_disco_m4, it fails
```
> west build -p auto -b  stm32h747i_disco_m4 samples/subsys/ipc/openamp/ 
-- west build: making build dir C:\zephyr\zephyr_project\zephyr\build pristine
-- west build: generating a build system
CMake Error at CMakeLists.txt:22 (message):
  stm32h747i_disco_m4 was not supported for this sample
```

So can you detail me how to build, flash and run this sample for both CM7 and CM4 ?

**Comment by cameled** [comment]
> > > > ow is my test result:
> > > > ```
> > > >   CM7:
> > > >      open-amp sample  -- working
> > > > 
> > > >   CM4:
> > > >     1. open-amp sample  -- not working
> > > >     2. blinky  ---working
> > > > ```
> > > 
> > > 
> > > Thanks for the tests and feedback. Though we should keep in might this may be related to other issues.
> > > What are openamp test result when BCM4 disabled ?
> > 
> > 
> > openamp work fine on CM4 when BCM4 disabled.
> 
> @cameled , I tried to reproduce your problem with openamp, but when I compile for stm32h747i_disco_m4, it fails
> 
> ```
> > west build -p auto -b  stm32h747i_disco_m4 samples/subsys/ipc/openamp/ 
> -- west build: making build dir C:\zephyr\zephyr_project\zephyr\build pristine
> -- west build: generating a build system
> CMake Error at CMakeLists.txt:22 (message):
>   stm32h747i_disco_m4 was not supported for this sample
> ```
> 
> So can you detail me how to build, flash and run this sample for both CM7 and CM4 ?

```
# Build M7 and M4
$ west build -p -b stm32h747i_disco_m7 samples/subsys/ipc/openamp

# flash M7
$ west flash

# flash M4 
$ west flash --build-dir build/openamp_remote-prefix/src/openamp_remote-build
```

**Comment by ABOSTM** [comment]
@cameled ,
I cherry picked this patch on top of your PR 38322, and samples/subsys/ipc/openamp is working fine on both CM7 and CM4.
(BCM4 enabled and disabled).

CM7 console:
```
*** Booting Zephyr OS build v2.7.0-rc1-99-gd3683888d740  ***
Starting application thread!

OpenAMP[master] demo started
Master core received a message: 1
Master core received a message: 3
Master core received a message: 5
Master core received a message: 7
Master core received a message: 9
Master core received a message: 11
Master core received a message: 13
Master core received a message: 15
Master core received a message: 17
Master core received a message: 19
Master core received a message: 21
Master core received a message: 23
Master core received a message: 25
Master core received a message: 27
Master core received a message: 29
Master core received a message: 31
Master core received a message: 33
Master core received a message: 35
Master core received a message: 37
Master core received a message: 39
Master core received a message: 41
Master core received a message: 43
Master core received a message: 45
Master core received a message: 47
Master core received a message: 49
Master core received a message: 51
Master core received a message: 53
Master core received a message: 55
Master core received a message: 57
Master core received a message: 59
Master core received a message: 61
Master core received a message: 63
Master core received a message: 65
Master core received a message: 67
Master core received a message: 69
Master core received a message: 71
Master core received a message: 73
Master core received a message: 75
Master core received a message: 77
Master core received a message: 79
Master core received a message: 81
Master core received a message: 83
Master core received a message: 85
Master core received a message: 87
Master core received a message: 89
Master core received a message: 91
Master core received a message: 93
Master core received a message: 95
Master core received a message: 97
Master core received a message: 99
OpenAMP demo ended.
```

Can you double check on your side ?

**Comment by cameled** [comment]
> @cameled ,
> I cherry picked this patch on top of your PR 38322, and samples/subsys/ipc/openamp is working fine on both CM7 and CM4.
> (BCM4 enabled and disabled).
> 
> CM7 console:
> 
> ```
> *** Booting Zephyr OS build v2.7.0-rc1-99-gd3683888d740  ***
> Starting application thread!
> 
> OpenAMP[master] demo started
> Master core received a message: 1
> Master core received a message: 3
> Master core received a message: 5
> Master core received a message: 7
> Master core received a message: 9
> Master core received a message: 11
> Master core received a message: 13
> Master core received a message: 15
> Master core received a message: 17
> Master core received a message: 19
> Master core received a message: 21
> Master core received a message: 23
> Master core received a message: 25
> Master core received a message: 27
> Master core received a message: 29
> Master core received a message: 31
> Master core received a message: 33
> Master core received a message: 35
> Master core received a message: 37
> Master core received a message: 39
> Master core received a message: 41
> Master core received a message: 43
> Master core received a message: 45
> Master core received a message: 47
> Master core received a message: 49
> Master core received a message: 51
> Master core received a message: 53
> Master core received a message: 55
> Master core received a message: 57
> Master core received a message: 59
> Master core received a message: 61
> Master core received a message: 63
> Master core received a message: 65
> Master core received a message: 67
> Master core received a message: 69
> Master core received a message: 71
> Master core received a message: 73
> Master core received a message: 75
> Master core received a message: 77
> Master core received a message: 79
> Master core received a message: 81
> Master core received a message: 83
> Master core received a message: 85
> Master core received a message: 87
> Master core received a message: 89
> Master core received a message: 91
> Master core received a message: 93
> Master core received a message: 95
> Master core received a message: 97
> Master core received a message: 99
> OpenAMP demo ended.
> ```
> 
> Can you double check on your side ?

Yes, you are right. I re-run the test on 3 stm32h747i_disco board, CM4 work fine.
Looks I do a bad test last day, sorry about that.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@cameled if satisfied, would you mind approving the PR, so it could get merged?

**Comment by cameled** [comment]
@erwango Yes, this PR can help to improve M4 boot logic.

**Comment by erwango** [comment]
> @erwango Yes, this PR can help to improve M4 boot logic.

I meant reviewing and approving (you don't need specific rights for that, normally)

**Review by cameled — APPROVED** [review]
LGTM.

**Review by cfriedt — APPROVED** [review]

**Pull request merged**